20251108-RIS-workflow-purge-ttl.md

```diff
@@ -0,0 +1,162 @@
+# Workflow: History Purge TTL on Completion
+
+* Author(s): @joshvanl
+
+## Overview
+
+This proposal details new functionality to the workflow runtime to give users the ability to delete completed workflow state from the actor state store after some configured time.
+All workflow instances may be configured with a unique TTL at workflow scheduling time.
+The default remains that workflow state will _not_ be deleted from the actor state store, and will remain there indefinitely.
+
+## Background
+
+It is currently the case that in order for users to delete old workflow state from the actor state store database, they either need to use the Purge Workflow API, or delete state from the database directly, either via out of Dapr database operations, or via using some kind of first class TTL feature of that database.
+Users typically want to delete old workflow state after some period of time from when the workflow has reached a terminal state.
+
+https://github.com/dapr/dapr/issues/9020
+
+## Design
+
+When scheduling a workflow, users will be able to configure some duration which upon elapsing after the workflow has reached a terminal state, the workflow will be purged from the actor state store.
+The duration will only start once the workflow has reached either a TERMINATED, COMPLETED, or FAILED state.
+
+Any duration may be given, i.e. days, weeks, or years.
+A duration of `0` may also be given, if the workflow actor state is wished to be deleted immediately after reaching a terminal state.
+
+### Usage
+
+#### CLI
+
+Users can give a Go style duration string when running a workflow from the CLI.
+
+```bash
+$ dapr run my-workflow --purge-ttl=5d
+```
+
+```bash
+$ dapr run my-workflow --purge-ttl=0s
+```
+
+The new purge reminders will be displayed like:
+
+```bash
+$ dapr scheduler list
+NAME                        BEGIN  COUNT  LAST TRIGGER
+purge-workflow/my-workflow  96h    0
+```
+
+#### Go
+
+```go
+wf.ScheduleWorkflow(ctx, "my-workflow", workflow.WithPurgeTTL(time.Hour*24*5))
```

Comment:
I read "purgeTTL" paired with 'ScheduleWorkflow" and my first thought was to scroll back up and validate that it only purges after the workflow hits a terminal state (it does), but without looking at documentation to confirm this, I fear the naming might be misleading and inadvertently convey that the TTL starts when the workflow is scheduled.

It's a little longer (but IDE autocomplete can trivially mitigate this), but might I suggest we instead call it something more self-documenting like "terminalPurgeTTL" or "postCompletionStateTTL" so it's clearer that the TTL window is started following a terminal state?